pkg/controller/lookup_cache.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"hash/adler32"
+	"sync"
+
+	"github.com/golang/groupcache/lru"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	hashutil "k8s.io/kubernetes/pkg/util/hash"
+)
+
+const maxCacheEntries = 4096
+
+// podKeyFunc return the key of a Pod, which is used to look up in the cache for it's matching ReplicationController/Job/DaemonSet.
+// Since we match Pod and ReplicationController/Job/DaemonSet by namespace and Labels/Selector, so if two Pods have the same
+// namespace and labels, they will have the same key.
+func podKeyFunc(pod *api.Pod) uint64 {
+	hash := adler32.New()
+	hashutil.DeepHashObject(hash, &equivalencePod{
+		namspace: pod.Namespace,
+		labels:   pod.Labels,
+	})
+	return uint64(hash.Sum32())
+}
+
+type equivalencePod struct {
+	namspace string
+	labels   map[string]string
+}
+
+// ReplicationControllerCache save the pod and replication controller matching relationship
+type ReplicationControllerCache struct {
```

Comment:
Similar comment. Maybe we can make this cache generic by having the value in a cache being runtime.Object?

Thanks to it you will avoid creating ReplicationControllerCache, ReplicaSetControllerCache, .... and you will have only one structure?